### PR TITLE
std: Fix linking against `signal` on Android

### LINKED ISSUE #236: Android linking fails with NDK r11b: undefined reference to 'bsd_signal'
Cross-compiling on Linux x64 to Android, rustc compiled from nightly sources as of 2016-03-17.

$ rustc --version -v
rustc 1.9.0-dev
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.9.0-dev

$ cat - >hello.rs
fn main() {
  println!("Hello, rusty Android!");
}
^D

$ rustc --target=arm-linux-androideabi -C linker=$HOME/Work/Android/NdkStandalone/bin/arm-linux-androideabi-gcc -C link-args="-pie -fPIE" hello.rs
error: linking with `/home/skirmantas/Work/Android/NdkStandalone/bin/arm-linux-androideabi-gcc` failed: exit code: 1
note: "/home/skirmantas/Work/Android/NdkStandalone/bin/arm-linux-androideabi-gcc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-Wl,--allow-multiple-definition" "-L" "/usr/local/stow/rustc-20160317/lib/rustlib/arm-linux-androideabi/lib" "hello.0.o" "-o" "hello" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "/usr/local/stow/rustc-20160317/lib/rustlib/arm-linux-androideabi/lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "/usr/local/stow/rustc-20160317/lib/rustlib/arm-linux-androideabi/lib/libstd-9026086f.rlib" "/usr/local/stow/rustc-20160317/lib/rustlib/arm-linux-androideabi/lib/libcollections-9026086f.rlib" "/usr/local/stow/rustc-20160317/lib/rustlib/arm-linux-androideabi/lib/librustc_unicode-9026086f.rlib" "/usr/local/stow/rustc-20160317/lib/rustlib/arm-linux-androideabi/lib/librand-9026086f.rlib" "/usr/local/stow/rustc-20160317/lib/rustlib/arm-linux-androideabi/lib/liballoc-9026086f.rlib" "/usr/local/stow/rustc-20160317/lib/rustlib/arm-linux-androideabi/lib/liballoc_jemalloc-9026086f.rlib" "/usr/local/stow/rustc-20160317/lib/rustlib/arm-linux-androideabi/lib/liblibc-9026086f.rlib" "/usr/local/stow/rustc-20160317/lib/rustlib/arm-linux-androideabi/lib/libcore-9026086f.rlib" "-l" "dl" "-l" "log" "-l" "gcc" "-l" "gcc" "-l" "c" "-l" "m" "-pie" "-fPIE" "-l" "compiler-rt"
note: /usr/local/stow/rustc-20160317/lib/rustlib/arm-linux-androideabi/lib/libstd-9026086f.rlib(std-9026086f.0.o):std.0.rs:function rt::lang_start::h1deaaba938bd63a9Xqz: error: undefined reference to 'bsd_signal'
collect2: error: ld returned 1 exit status

error: aborting due to previous error

### PR DESCRIPTION
**Pull request by alexcrichton**
Currently the minimum supported Android version of the standard library is
API level 18 (android-18). Back in those days [1](https://chromium.googlesource.com/android_tools/+/20ee6d20/ndk/platforms/android-18/arch-arm/usr/include/signal.h) the `signal` function was
just an inline wrapper around `bsd_signal`, but starting in API level
android-20 the `signal` symbols was introduced [2](https://chromium.googlesource.com/android_tools/+/fbd420/ndk_experimental/platforms/android-20/arch-arm/usr/include/signal.h). Finally, in android-21
the API `bsd_signal` was removed [3](https://chromium.googlesource.com/android_tools/+/20ee6d/ndk/platforms/android-21/arch-arm/usr/include/signal.h).

Basically this means that if we want to be binary compatible with multiple
Android releases (oldest being 18 and newest being 21) then we need to check
for both symbols and not actually link against either.

This was first discovered in rust-lang/libc#236 with a fix proposed in
rust-lang/libc#237. I suspect that we'll want to accept rust-lang/libc#237 so
Rust crates at large continue to be compatible with newer releases of Android
and crates, like the standard library, that want to opt into older support can
continue to do so via similar means.

Closes rust-lang/libc#236

### CODE REVIEW

**Comment by alexcrichton** [comment]
Looks like `weak!` isn't working on Android, both symbols are returning `None` and then we're panicking on startup.

Gee I wonder why `weak!` isn't working? Could it have something to do with `dlopen(NULL, RTLD_LAZY)`? Could it have perhaps been [previously reported even](https://github.com/rust-lang/rust/issues/10356)?!

Ok, so that strategy doesn't work on Android, but apparently `RTLD_DEFAULT` does as a direct argument to `dlsym`, so just [gotta add it to libc](https://github.com/rust-lang/libc/pull/254)

**Comment by alexcrichton** [comment]
Hm this may end up being difficult to land and have ramifications on the rest of the Rust ecosystem. This updates the libc submodule to include changes in https://github.com/rust-lang/rfcs/pull/1529 which is starting to link a new library by default. I'm wary of this causing breakage on bots beyond ours, so we should keep our eyes peeled and be ready to revert if need be.
